datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java

```diff
@@ -141,6 +141,116 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
+   * @param partitionsMetadata the subscribed partitions metadata received from connector
+   * @return new assignment
+   */
+  public Map<String, Set<DatastreamTask>> movePartitions(Map<String, Set<DatastreamTask>> currentAssignment,
+      Map<String, Set<String>> targetAssignment, DatastreamGroupPartitionsMetadata partitionsMetadata) {
+
+    LOG.info("Move partition, task: {}, target assignment: {}, all partitions: {}", currentAssignment,
+        targetAssignment, partitionsMetadata.getPartitions());
+
+    String dgName = partitionsMetadata.getDatastreamGroup().getName();
+
+    Set<String> allToReassignPartitions = new HashSet<>();
+    targetAssignment.values().stream().forEach(allToReassignPartitions::addAll);
+    allToReassignPartitions.retainAll(partitionsMetadata.getPartitions());
+
+    //construct a map to store the tasks and if it contain the partitions that need to be released
+    //map: <taskName, partitions that need to be released>
+    Map<String, Set<String>> toReleasePartitionsTaskMap = new HashMap<>();
+    Map<String, String> partitionToSourceTaskMap = new HashMap<>();
+
+    //We first confirmed that the partitions in the target assignment can be removed, and we found its source task
+    currentAssignment.keySet().stream().forEach(instance -> {
+      Set<DatastreamTask> tasks = currentAssignment.get(instance);
+      tasks.forEach(task -> {
+        if (dgName.equals(task.getTaskPrefix())) {
+          Set<String> toMovePartitions = new HashSet<>(task.getPartitionsV2());
+          toMovePartitions.retainAll(allToReassignPartitions);
+          toReleasePartitionsTaskMap.put(task.getDatastreamTaskName(), toMovePartitions);
+          toMovePartitions.forEach(p -> partitionToSourceTaskMap.put(p, task.getDatastreamTaskName()));
+        }
+      });
+    });
+
+    Set<String> tasksToMutate = toReleasePartitionsTaskMap.keySet();
+    Set<String> toReleasePartitions = new HashSet<>();
+    toReleasePartitionsTaskMap.values().forEach(v -> toReleasePartitions.addAll(v));
+
+    //Compute new assignment from the current assignment
+    Map<String, Set<DatastreamTask>> newAssignment = new HashMap<>();
+
+    currentAssignment.keySet().stream().forEach(instance -> {
+      Set<DatastreamTask> tasks = currentAssignment.get(instance);
+
+      // check if this instance has any partition to be moved
+      final Set<String> toMovedPartitions = new HashSet<>();
+      if (targetAssignment.containsKey(instance)) {
+        toMovedPartitions.addAll(targetAssignment.get(instance).stream().filter(toReleasePartitions::contains).collect(Collectors.toSet()));
+      }
+
+      Set<DatastreamTask> dgTasks = tasks.stream().filter(t -> dgName.equals(t.getTaskPrefix()))
+          .collect(Collectors.toSet());
+      if (toMovedPartitions.size() > 0 && dgTasks.isEmpty()) {
+        throw new DatastreamRuntimeException("No task is available in the target instance " + instance);
```

Comment:
How do these errors get communicated to the operator that partition movement failed because of this issue.

Reply:
Added a metric for num_errors in Coordinator where caught this exception.